lib/spack/spack/test/multimethod.py

```diff
@@ -25,142 +26,83 @@ def pkg_name(request):
-    pkg = spack.repo.get(pkg_name + '@1.0')
-    assert pkg.no_version_2() == 1
-
-    pkg = spack.repo.get(pkg_name + '@3.0')
-    assert pkg.no_version_2() == 3
-
-    pkg = spack.repo.get(pkg_name + '@4.0')
-    assert pkg.no_version_2() == 4
-
-
-def test_version_overlap(pkg_name):
-    pkg = spack.repo.get(pkg_name + '@2.0')
-    assert pkg.version_overlap() == 1
-
-    pkg = spack.repo.get(pkg_name + '@5.0')
-    assert pkg.version_overlap() == 2
-
-
-def test_mpi_version(pkg_name):
-    pkg = spack.repo.get(pkg_name + '^mpich@3.0.4')
-    assert pkg.mpi_version() == 3
-
-    pkg = spack.repo.get(pkg_name + '^mpich2@1.2')
-    assert pkg.mpi_version() == 2
-
-    pkg = spack.repo.get(pkg_name + '^mpich@1.0')
-    assert pkg.mpi_version() == 1
-
-
-def test_undefined_mpi_version(pkg_name):
-    pkg = spack.repo.get(pkg_name + '^mpich@0.4')
-    assert pkg.mpi_version() == 1
-
-    pkg = spack.repo.get(pkg_name + '^mpich@1.4')
-    assert pkg.mpi_version() == 1
-
-
-def test_default_works(pkg_name):
-    pkg = spack.repo.get(pkg_name + '%gcc')
-    assert pkg.has_a_default() == 'gcc'
-
-    pkg = spack.repo.get(pkg_name + '%intel')
-    assert pkg.has_a_default() == 'intel'
-
-    pkg = spack.repo.get(pkg_name + '%pgi')
-    assert pkg.has_a_default() == 'default'
+        spec.package.no_version_2()
+
+
+@pytest.mark.parametrize('constraint_str,method_name,expected_result', [
+    # Only one version match these constraints
+    ('@1.0', 'no_version_2', 1),
+    ('@3.0', 'no_version_2', 3),
+    ('@4.0', 'no_version_2', 4),
+    # These constraints overlap, in which case the first match wins
+    ('@2.0', 'version_overlap', 1),
+    ('@5.0', 'version_overlap', 2),
+    # These constraints are on the version of a virtual dependency
+    ('^mpich@3.0.4', 'mpi_version', 3),
+    ('^mpich2@1.2', 'mpi_version', 2),
+    ('^mpich@1.0', 'mpi_version', 1),
+    # Undefined mpi versions
+    ('^mpich@0.4', 'mpi_version', 1),
+    ('^mpich@1.4', 'mpi_version', 1),
+    # Constraints on compilers with a default
+    ('%gcc', 'has_a_default', 'gcc'),
+    ('%intel', 'has_a_default', 'intel'),
+    ('%pgi', 'has_a_default', 'default'),
+    # Constraints on dependencies
+    ('^zmpi', 'different_by_dep', 'zmpi'),
+    ('^mpich', 'different_by_dep', 'mpich'),
+    # If we try to switch on some entirely different dep, it's ambiguous,
+    # but should take the first option
+    ('^foobar', 'different_by_dep', 'mpich'),
+    # Constraints on virtual dependencies
+    ('^mpich2', 'different_by_virtual_dep', 2),
+    ('^mpich@1.0', 'different_by_virtual_dep', 1),
+    # Multimethod with base classes
+    ('@1', 'base_method', 'base_method'),
+    # Boolean
+    ('', 'boolean_true_first', 'True'),
+    ('', 'boolean_false_first', 'True')
+])
+def test_multimethod_calls(pkg_name, constraint_str, method_name, expected_result):
+    s = spack.spec.Spec(pkg_name + constraint_str)
+    assert getattr(s.package, method_name)() == expected_result
```

Comment:
I dunno if I like this -- it was clearer before and the test method names would at least give you better error messages from `pytest`.  Is there a reason other than style to parameterize like this?

Reply:
The main reason for me was to avoid code duplication. All these tests are checking if a multimethod is called an expected number of times under different conditions:
1. Before, to add a test, we had a cut and paste of the same code, slightly modified for a few strings
2. Now, we just need to add a new set of parameters

If it was structured like it is now, it would have been a single line change in the code, instead of 10-20.

Reply:
I kinda agree with both points, maybe add a custom assert message that re-creates the more useful output we would otherwise be getting from the method names?